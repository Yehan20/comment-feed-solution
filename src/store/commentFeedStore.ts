import { defineStore } from 'pinia';
import type { LocalStorage, Message } from '../types/message';
import { helperDownvote, helperUpvote, helperLocalStorage, helperUpdateLocalStorage } from '../utils/messageHelper';
import { questionFeeds } from '../data/questionFeeds';
import { QuestionFeed } from '../types/questionfeed';
import Images from '../assets/images/profileImages';

/*
   Upvote | Down Vote Logic based on reddit

   updvote click increase by one
   updvote unlick click derease by one

   downvote click decrease by one
   downvote unlick increase by one

   upvote  clicked and down vote clicked decrease by two
   downvote clicked and upvote clicked increase by two

*/



//Code of the Store 
export const usecommentFeedStore = defineStore('comments', {

    state: () => {
        return {

            commentFeed:  [

                {
                    id: 1,
                    message: "from azerbaijan",
                    parentId: null,
                    points: 7,
                    isUpvoted: false,
                    isDownvoted: true,
                    userName: "Zeus",
                    profilePic: Images.Zeus,
                    createdAt: new Date("2021-11-08")
                    ,
            
                },
                {
                    id: 2,
                    message: "hello world",
                    parentId: 1,
                    points: 3,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Hades",
                    profilePic: Images.Hades,
                    createdAt: new Date("2024-10-08")
                    ,
            
                },
            
                {
                    id: 111,
                    message: "hello world",
                    parentId: 1,
                    points: 3,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Hades ge malli",
                    profilePic: Images.Hades,
                    createdAt: new Date("2024-10-08")
                    ,
            
                },
                {
                    id: 3,
                    message: "hello world",
                    parentId: 2,
                    points: 5,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Hestia",
                    profilePic: Images.Hestia,
                    createdAt: new Date("2024-11-01"),
            
                },
                {
                    id: 4,
                    message: "hello world",
                    parentId: 3,
                    points: -1,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Appollo",
                    profilePic: Images.Appollo,
                    createdAt: new Date("2024-08-08"),
            
                },
                {
                    id: 5,
                    message: "hello world",
                    parentId: 4,
                    points: 10,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Athena",
                    profilePic: Images.Athena,
                    createdAt: new Date("2024-05-08"),
            
                },
                {
                    id: 6,
                    message: "hello world",
                    parentId: 5,
                    points: 5,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Kronos",
                    profilePic: Images.Kronos,
                    createdAt: new Date("2022-11-08"),
            
                },
                {
                    id: 7,
                    message: "hello world",
                    parentId: 6,
                    points: 7,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Ariest",
                    profilePic: Images.Aries,
                    createdAt: new Date("2023-11-08"),
            
                },
            
            
            
            
                {
                    id: 8,
                    message: "Yehan",
                    parentId: null,
                    points: 8,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Dog",
                    profilePic: Images.Aries,
                    createdAt: new Date("2023-11-08"),
            
                },
            
            
            
                {
                    id: 9,
                    message: "Nilnaga",
                    parentId: null,
                    points: 9,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Cat",
                    profilePic: Images.Aries,
                    createdAt: new Date("2023-11-08"),
            
                },
            
                {
                    id: 10,
                    message: "Don",
                    parentId: null,
                    points: 10,
                    isUpvoted: false,
                    isDownvoted: false,
                    userName: "Lion",
                    profilePic: Images.Aries,
                    createdAt: new Date("2023-11-08"),
            
                },
            
            ],
            isSortByUpvote: false,
            isLoaded: false,
            totalPoints: 0,
            isUpvoted: false,
            isDownvoted: false,
            askedBy: "",
            questionId: 'Q1',
            question: '',
            questionDesc: "",
            mounted: false,
        }
    },

    getters: {

        getComments(state) {

            if (state.isSortByUpvote) {
                return (id: number | null) => state.commentFeed.filter((comment) => comment.parentId === id).sort((a, b) => b.points - a.points);
            }
            return (id: number | null) => state.commentFeed.filter((comment) => comment.parentId === id)
        },

        getReplies(state) {
      
            if (state.isSortByUpvote) {

                return state.commentFeed.filter((comment) => comment.parentId !== null).sort((a, b) => b.points - a.points);

            }
            return state.commentFeed.filter((replyForComment) => replyForComment.parentId !== null); // ld    
        },

        getRepliesPerComment() {



            return (id: number) => this.getReplies.filter((reply) => reply.parentId == id)
        },

        allRepliesPerComment() {
            return (id: number) => this.getReplies.filter((reply) => reply.id === id)
        },


    },

    actions: {

        addComment(comment: Message) {

            this.commentFeed.push(comment);

            helperUpdateLocalStorage(this.questionId, { commentFeed: [...this.commentFeed] })

        },

        upovoteMain() {

            if (!this.isDownvoted && !this.isUpvoted) { // comment  not upovted or not downvoted

                this.totalPoints = this.totalPoints + 1
                this.isUpvoted = !this.isUpvoted

                helperUpdateLocalStorage(this.questionId, {
                    isDownvoted: this.isDownvoted,
                    isUpvoted: this.isUpvoted,
                    totalPoints: this.totalPoints
                })
               return;
            }

            if (!this.isDownvoted && this.isUpvoted) {  // comment is upvoted

                this.totalPoints = this.totalPoints - 1
                this.isUpvoted = !this.isUpvoted

                helperUpdateLocalStorage(this.questionId, {
                    isDownvoted: this.isDownvoted,
                    isUpvoted: this.isUpvoted,
                    totalPoints: this.totalPoints
                })
                return;
            }

            if (this.isDownvoted && !this.isUpvoted) {  // comment is downvoted not upvoted

                this.totalPoints += 2; // change 
                this.isUpvoted = !this.isUpvoted; // same
                this.isDownvoted = !this.isDownvoted;

                helperUpdateLocalStorage(this.questionId, {
                    isDownvoted: this.isDownvoted,
                    isUpvoted: this.isUpvoted,
                    totalPoints: this.totalPoints
                })
               return;
            }

        },

        upvoteComment(id: number) {


   
            let changedComment = this.commentFeed.find((comment) => comment.id === id) as Message
            helperUpvote(changedComment);

            helperUpdateLocalStorage(this.questionId, { commentFeed: [...this.commentFeed] })

        },

        downvoteComment(id: number) {

            let changedComment = this.commentFeed.find((comment) => comment.id === id) as Message
            helperDownvote(changedComment);

            helperUpdateLocalStorage(this.questionId, { commentFeed: [...this.commentFeed] })
        },

        downvoteMain() {

            if (!this.isDownvoted && !this.isUpvoted) { // comment not donwvoted or upvoted

                this.totalPoints = this.totalPoints - 1
                this.isDownvoted = !this.isDownvoted
                helperUpdateLocalStorage(this.questionId, { isDownvoted: this.isDownvoted, isUpvoted: this.isUpvoted, totalPoints: this.totalPoints })
                return;

            }


            if (this.isDownvoted && !this.isUpvoted) {  // comment is downvoted and not upvoted 

                this.totalPoints = this.totalPoints + 1
                this.isDownvoted = !this.isDownvoted
                helperUpdateLocalStorage(this.questionId, { isDownvoted: this.isDownvoted, isUpvoted: this.isUpvoted, totalPoints: this.totalPoints })
                return;

            }

            if (!this.isDownvoted && this.isUpvoted) { // comment is not downvoted but upvoted

                this.totalPoints -= 2; // change 
                this.isUpvoted = !this.isUpvoted; // same
                this.isDownvoted = !this.isDownvoted;
                helperUpdateLocalStorage(this.questionId, { isDownvoted: this.isDownvoted, isUpvoted: this.isUpvoted, totalPoints: this.totalPoints })

                return;

            }


        },

        sortByUpvote() {

            if (!this.isSortByUpvote) {
                this.isSortByUpvote = true
                this.isLoaded = false;

                helperUpdateLocalStorage(this.questionId, { isSortByUpvote: this.isSortByUpvote })
            }


        },

        loadFeed() {

            const fromLocal = helperLocalStorage(this.questionId) as LocalStorage | null;

            if (!fromLocal) {
                const questionFeed: QuestionFeed = {...questionFeeds.find((questionFeed) => questionFeed.id === this.questionId) as QuestionFeed};


                this.commentFeed = questionFeed.commentFeed
                this.totalPoints = questionFeed.points;
                this.askedBy = questionFeed.askedBy;
                this.isDownvoted = questionFeed.isDownvoted;
                this.isUpvoted = questionFeed.isUpvoted;
                this.question = questionFeed.question;
                this.questionDesc = questionFeed.questionDesc


                helperUpdateLocalStorage(this.questionId,
                    {
                        commentFeed: this.commentFeed,
                        isSortByUpvote: this.isSortByUpvote,
                        askedBy: this.askedBy,
                        question: this.question,
                        questionDesc: this.questionDesc,
                        isDownvoted: this.isDownvoted,
                        isUpvoted: this.isUpvoted,
                        totalPoints: this.totalPoints

                    })

            }
            else {

                this.commentFeed = fromLocal.commentFeed as Message[];
                this.isSortByUpvote = fromLocal.isSortByUpvote as boolean;
                this.askedBy = fromLocal.askedBy as string;
                this.question = fromLocal.question as string;
                this.questionDesc = fromLocal.questionDesc as string;
                this.isDownvoted = fromLocal.isDownvoted as boolean;
                this.isUpvoted = fromLocal.isUpvoted as boolean;
                this.totalPoints = fromLocal.totalPoints as number;
            }

            this.isLoaded = true;

        },

        resetSort() {

            if (this.isSortByUpvote) {
                this.isSortByUpvote = false
                this.isLoaded = false;
            }


            helperUpdateLocalStorage('Q1', { isSortByUpvote: this.isSortByUpvote })
        },

        changeQuestion(qId: string) {
           
  
            this.questionId = qId;
            this.loadFeed();
      
        }


    }


})

