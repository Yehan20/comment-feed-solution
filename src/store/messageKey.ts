import { InjectionKey, Ref } from "vue";

export const Key =Symbol() as InjectionKey<Ref<number | null >>;