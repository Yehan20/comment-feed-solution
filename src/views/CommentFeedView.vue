<script setup lang="ts">
import CommentSection from '../components/CommentSection.vue'
import TextEditor from '../components/TextEditor.vue';
import { usecommentFeedStore } from '../store/commentFeedStore';
  
const commentFeedStore = usecommentFeedStore();

</script>
<template>
    <div class="comment__feed">


        <div class="question__wrapper">

            <h1>What came first the Egg or the Hen ??</h1>
            <p>{{commentFeedStore.text}}</p>

            <div class="question__detail">
                 <p>I have been wondering what came first some says egg some says hen  
                    my Teacher asked it told me to come up with the answer tommarow even which answer i have i need to prove it 
                    any thoughts?
                </p>
            </div>
            <div class="question__stats">
                 <!-- Upvote down vote List -->
            </div>
        </div>
        <TextEditor :showDiscussion="true"/>
        <CommentSection/>
    </div>
</template>


<style lang="scss" scoped>
  @use '../assets/scss/mixins.scss' as m;
  
  .comment__feed {
    @include m.verticalPadding(40px,40px);
  }
  .question__wrapper {
      @include m.verticalMargin(0px,20px)
  }
  .question__detail {
     max-width: 750px;
     line-height: 1.6;
  }
 
</style>