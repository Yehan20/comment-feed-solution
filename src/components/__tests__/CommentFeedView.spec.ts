import { describe, it, expect, beforeEach } from "vitest";
import {  mount } from "@vue/test-utils";
import { createPinia, setActivePinia } from "pinia";
import CommentFeedView from "../../views/CommentFeedView.vue";
import { useCommentFeedStore } from "../../store/commentFeedStore";
import { addIcons, OhVueIcon } from "oh-vue-icons";
import { nextTick } from "vue";
import { RiLoader2Line } from "oh-vue-icons/icons";


// Test for Comment View Component
describe('ComentFeedView', () => {


    beforeEach(() => {
        setActivePinia(createPinia());


    });


    it("Should Render Correctly", async () => {
        const wrapper = mount(CommentFeedView, {
            global: {
                components: {
                    'v-icon': OhVueIcon,
                },
            },
        });

        // Wait for the promises to resolve (if any)
        addIcons(RiLoader2Line);


        // Then perform the snapshot test
        expect(wrapper.html()).toMatchSnapshot();
    });

    // Sort by Upvote Click Event Test
    it('Should Change the isSorted State', async () => {
        const commentFeedStore = useCommentFeedStore(); // Get the store instance
      

        // Mount the component with the store
        const wrapper = mount(CommentFeedView, {

            global: {
                components: {
                    'v-icon': OhVueIcon,  // Register v-icon manually in test
                },
            },

        });

        const button = wrapper.find('.btn__sort');
        await button.trigger('click'); // will change the sort state to true in our 


        await wrapper.vm.$nextTick();

        expect(commentFeedStore.isSortByUpvote).toBe(true)
    });


    // Reset Click Event Test
    it('Should Reset  the isSorted State Change', async () => {

        const commentFeedStore = useCommentFeedStore(); // Get the store instance

        // Mount the component with the store
        const wrapper = mount(CommentFeedView, {

            global: {
                components: {
                    'v-icon': OhVueIcon,  // Register v-icon manually in test
                },
            },

        });


        const button = wrapper.find('.btn__reset');
        await button.trigger('click'); // will change the sort state to tru in our 


        await wrapper.vm.$nextTick();


        expect(commentFeedStore.isSortByUpvote).toBe(false)
    });

    // Upvote Main Points Test
    it("should Upvote and increase Main Points ", async () => {

        const wrapper = mount(CommentFeedView, {

            global: {
                components: {
                    'v-icon': OhVueIcon,  // Register v-icon manually in test
                },
            },
        });
        const commentFeedStore = useCommentFeedStore(); // Get the store instance



        const button = wrapper.findComponent({ name: 'CommentVoteButtonGroup' }).find('.btn__upvote');
        await button.trigger('click');

        await nextTick();

        // Check if the emit was triggered with the correct value
        const emittedEvents = wrapper.findComponent({ name: 'CommentVoteButtonGroup' }).emitted('@upVote');
        expect(emittedEvents).toBeTruthy();
        expect(commentFeedStore.totalPoints).toEqual(1)
    });


    // Downvote Main Points Test
    it("should Downvote and Decrease Main Points ", async () => {

        const wrapper = mount(CommentFeedView, {

            global: {
                components: {
                    'v-icon': OhVueIcon,  // Register v-icon manually in test
                },
            },
        });
        const commentFeedStore = useCommentFeedStore(); // Get the store instance



        const button = wrapper.findComponent({ name: 'CommentVoteButtonGroup' }).find('.btn__downvote');
        await button.trigger('click');

        await nextTick();

        // Check if the emit was triggered with the correct value
        const emittedEvents = wrapper.findComponent({ name: 'CommentVoteButtonGroup' }).emitted('@downVote');
        expect(emittedEvents).toBeTruthy();
        expect(commentFeedStore.totalPoints).toEqual(-1)
    });

})
