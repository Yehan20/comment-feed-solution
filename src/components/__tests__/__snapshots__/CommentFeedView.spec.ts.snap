// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ComentFeedView > Should Render Correctly 1`] = `
"<div data-v-fd6b37a4="" class="comment__feed">
  <div data-v-fd6b37a4="" class="question__wrapper">
    <h1 data-v-fd6b37a4=""></h1>
    <div data-v-fd6b37a4="" class="question__detail">
      <h2 data-v-fd6b37a4=""></h2>
    </div>
    <div data-v-fd6b37a4="" class="question__stats">
      <!-- Upvote down vote List -->
      <div data-v-fd6b37a4="" class="bg-white rounded"><button data-v-1090a549="" class="btn btn__upvote" title="Click to Upvote"><svg data-v-1090a549="" class="ov-icon" style="font-size: 1.44em;" aria-hidden="true" width="23.04" height="23.04" viewBox="-1.6 -1.6 19.2 19.2" fill="currentColor">
            <path fill-rule="evenodd" d="M1 8a7 7 0 1014 0A7 7 0 001 8zm15 0A8 8 0 110 8a8 8 0 0116 0zm-7.5 3.5a.5.5 0 01-1 0V5.707L5.354 7.854a.5.5 0 11-.708-.708l3-3a.5.5 0 01.708 0l3 3a.5.5 0 01-.708.708L8.5 5.707V11.5z"></path>
          </svg></button><button data-v-1090a549="" class="btn btn__downvote" title="Click to DownVote"><svg data-v-1090a549="" class="ov-icon" style="font-size: 1.44em;" aria-hidden="true" width="23.04" height="23.04" viewBox="-1.6 -1.6 19.2 19.2" fill="currentColor">
            <path fill-rule="evenodd" d="M1 8a7 7 0 1014 0A7 7 0 001 8zm15 0A8 8 0 110 8a8 8 0 0116 0zM8.5 4.5a.5.5 0 00-1 0v5.793L5.354 8.146a.5.5 0 10-.708.708l3 3a.5.5 0 00.708 0l3-3a.5.5 0 00-.708-.708L8.5 10.293V4.5z"></path>
          </svg></button></div>
      <p data-v-fd6b37a4="" class="m-none">0 points</p>
    </div>
  </div>
  <div data-v-abc9304f="" data-v-fd6b37a4="" class="text__editor__container">
    <div data-v-abc9304f="" class="text__editor__top">
      <h3 data-v-abc9304f="">Discussion (11 comments)</h3>
      <!--v-if-->
    </div>
    <form data-v-abc9304f="" role="form"><textarea data-v-abc9304f="" name="" id="" rows="10" class="text__editor" placeholder="Make it Creative :)" value=""></textarea>
      <div data-v-abc9304f="" class="text__editor__footer">
        <p data-v-abc9304f="">0 / 400 Characters </p><button data-v-abc9304f="" class="btn btn__comment"> Comment </button>
      </div>
    </form>
    <!--v-if-->
  </div>
  <div data-v-fd6b37a4="" class="filter__wrapper"><button data-v-fd6b37a4="" title="Sort by Most Points" class="btn btn__sort"> Most Upvoted </button><button data-v-fd6b37a4="" title="Reset" class="selected btn btn__reset"> Reset </button></div>
  <div data-v-fd6b37a4="" class="loader__wraper"><svg data-v-fd6b37a4="" class="ov-icon" style="font-size: 2.4em;" aria-hidden="true" width="38.4" height="38.4" viewBox="0 0 24 24" fill="currentColor" color="red">
      <path fill="none" d="M0 0h24v24H0z"></path>
      <path d="M12 2a1 1 0 011 1v3a1 1 0 01-2 0V3a1 1 0 011-1zm0 15a1 1 0 011 1v3a1 1 0 01-2 0v-3a1 1 0 011-1zm10-5a1 1 0 01-1 1h-3a1 1 0 010-2h3a1 1 0 011 1zM7 12a1 1 0 01-1 1H3a1 1 0 010-2h3a1 1 0 011 1zm12.071 7.071a1 1 0 01-1.414 0l-2.121-2.121a1 1 0 011.414-1.414l2.121 2.12a1 1 0 010 1.415zM8.464 8.464a1 1 0 01-1.414 0l-2.12-2.12a1 1 0 011.414-1.415l2.12 2.121a1 1 0 010 1.414zM4.93 19.071a1 1 0 010-1.414l2.121-2.121a1 1 0 111.414 1.414l-2.12 2.121a1 1 0 01-1.415 0zM15.536 8.464a1 1 0 010-1.414l2.12-2.121a1 1 0 011.415 1.414L16.95 8.464a1 1 0 01-1.414 0z"></path>
    </svg></div>
</div>"
`;
